### PR TITLE
Initial SMT store type

### PR DESCRIPTION
**Pull request by tzdybal**
## Description

This PR introduces SMT store layer. We need to wait for conclusion of [ADR-040](https://github.com/cosmos/cosmos-sdk/pull/8430) before moving forward.

Current state implements most interfaces (except `Queryable`), and provides compatibility layer for Tendermint ProofOp's.

### CODE REVIEW

File: store/smt/iterator_test.go
Comment (tzdybal) [inline store/smt/iterator_test.go:15]: Table driven tests

File: store/smt/store.go
Comment (tzdybal) [inline store/smt/store.go:63]: SMT propagates errors from underlying storage, so we should panic (and probably log before)

File: store/smt/iterator.go
Comment (liamsi) [inline store/smt/iterator.go:17]: This isn't really self-explanatory. Add a brief explanation (even if it is unexported).

Nit: I'd prefer using `string` `const`ants instead and wrap them  with `[]byte()` where they are used.

File: store/smt/proof.go
Comment (liamsi) [inline store/smt/proof.go:69]:
```suggestion
		Type: SMTProofType,
```

**Review by liamsi — COMMENTED** [review]
Great start to get the SMT to adhere to the required interfaces 👍🏼

**Comment by liamsi** [comment]
> We need to wait for conclusion of ADR-040 before moving forward.

Does that mean we should hold off from merging as well?

**Comment by tzdybal** [comment]
> Does that mean we should hold off from merging as well?

Current state of implementation is a complete, usable store, with support for proofs. We can merge it, and then continue work, or we can just leave it as is, to be a starting point for further work.

File: store/smt/proof_test.go
Comment (robert-zaremba) [inline store/smt/proof_test.go:33]: instead of adding `t` to every assert or require I usually shadow the package names:
```
assert := assert.New(t)
require := require.New(t)
// and then:
assert.Equal(root, r[0])  // no need to pass `t`
```

Comment (tzdybal) [reply store/smt/proof_test.go:33]: Nice one!

**Review by liamsi — APPROVED** [review]
I'm OK with merging this in its current form into our fork. 

I would certainly hold-off from _replacing_ iavl with the smt untill we have a MVP without fraud-proofs. Maybe, we should even implement fraud proofs using iavl first.

**Comment by robert-zaremba** [comment]
I think the plan was to make a new PR and target it to `cosmos-sdk/store/ll-smt`  and then review.

**Comment by liamsi** [comment]
That's right! I actually forgot. But we could do both. I'll leave it up to @tzdybal to decide.

**Comment by robert-zaremba** [comment]
> That's right! I actually forgot. But we could do both. I'll leave it up to @tzdybal to decide.

What's the advantage of doing both? Wouldn't it complicate?

**Comment by liamsi** [comment]
It would only complicate things if we'd have to versions that we try to keep in sync, one in that branch and one in our fork.

**Comment by tzdybal** [comment]
Drafted https://github.com/cosmos/cosmos-sdk/pull/8507.

**Comment by adlerjohn** [comment]
What's the status on implementing the SMT here @tzdybal?

**Comment by adlerjohn** [comment]
Converting to draft but keeping around since we may need this branch soon

**Comment by adlerjohn** [comment]
cc @Manav-Aggarwal @jbowen93

**Comment by evan-forbes** [comment]
can we close this?

**Comment by evan-forbes** [comment]
closing as I think this is outdated

pls reopen or move if not slightly_smiling_face

**Pull request closed without merging**
